[SelectionDAG] Utilizing target hook convertSelectOfConstantsToMath for SelectwithConstant

vg0204 opened the pull request:
The Target hook convertSelectOfConstantsToMath() needs to be used within SimplifySelectCC helper combine function in SelectionDAG Isel, where generic select folding with constants is happening into simple maths op using the condition as it is.

It necessarily fixes #121145.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -28189,7 +28189,7 @@ SDValue DAGCombiner::SimplifySelectCC(const SDLoc &DL, SDValue N0, SDValue N1,
   bool Swap = N3C && isNullConstant(N2) && N3C->getAPIntValue().isPowerOf2();
 
-  if ((Fold || Swap) &&
+  if (TLI.convertSelectOfConstantsToMath(VT) && (Fold || Swap) &&
       TLI.getBooleanContents(CmpOpVT) ==
           TargetLowering::ZeroOrOneBooleanContent &&
       (!LegalOperations || TLI.isOperationLegal(ISD::SETCC, CmpOpVT))) {
```
```suggestion
  if ((Fold || Swap) &&
      TLI.getBooleanContents(CmpOpVT) ==
          TargetLowering::ZeroOrOneBooleanContent &&
      (!LegalOperations || TLI.isOperationLegal(ISD::SETCC, CmpOpVT)) &&
      TLI.convertSelectOfConstantsToMath(VT)) {
```

Order the complex checks last

arsenm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.h:
```diff
@@ -796,6 +796,8 @@ class AArch64TargetLowering : public TargetLowering {
   bool shouldConvertConstantLoadToIntImm(const APInt &Imm,
                                          Type *Ty) const override;
 
+  bool convertSelectOfConstantsToMath(EVT VT) const override { return true; }
```
I would do this as a separate PR, the current test diffs mix in the effect of the new call and the new override

vg0204 replied:
Point noted!

arsenm reviewed (commented):
Description should have fewer tags and name the used target hook

vg0204 commented on llvm/test/CodeGen/Thumb/branchless-cmp.ll:
```diff
@@ -1,3 +1,4 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
```
Sure!

vg0204 commented on llvm/test/CodeGen/Thumb/branchless-cmp.ll:
```diff
@@ -55,10 +56,12 @@ entry:
+; CHECK-NEXT:    beq .LBB4_2
+; CHECK-NEXT:  @ %bb.1:
+; CHECK-NEXT:    movs r0, #4
+; CHECK-NEXT:  .LBB4_2: @ %entry
```
> Is the end of the function getting cut off by update_llc_test_checks.py? If auto-generation isn't working correctly, please take a look.

Removed them expliclty after auto-generation, thought to only keep changed relevant things!

vg0204 commented:
@efriedma-quic , I am not quite sure about AArch64 tests, (as not much familiar with its target specific optimization). Even as @arsenm suggested the new override for target hook (should be taken care separately) mentioned for AArch64 apparently looked beneficial to me so did that based on changed LIT test after hook usgae. Can you comment even that override is needed for AArch64!

vg0204 committed:
[SelectionDAG] Added target hook check for SelectWithConstant

This patch adds the required convertSelectOfConstantsToMath() target
hook within SimplifySelectCC helper combine function in SelectionDAG
Isel, where generic select folding with constants is happening into
simple maths op using the condition as it is.

vg0204 commented:
Considered to add the override for any other target as separate PR, with proper reasoning there, & see their cascading effect with this PR separately!

efriedma-quic reviewed (approved):
LGTM

I suspect we might want an override for AArch64, and for Thumb1, after some further investigation.  But the differences are small enough I'm fine with taking this as-is, and leaving investigating those for later.
